### PR TITLE
5821 unit testing

### LINKED ISSUE #5821: Unit Test `formatIdsForSolrClause` of `SolrQueryFormatter.java`
Hi guys,

as described in #5619 I focus on unit testing. This time, I'll have a look at the function below.

Until now, the function was only tested indirectly and thus not completely (even though coveralls shows that most of the lines are covered, [see here](https://coveralls.io/builds/23241781/source?filename=src/main/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatter.java#L35)). More importantly, the function is not tested with considering different "input combinations". To directly call the function for proper testing, I need to change the method's access modifier from `private` to `protected`.

https://github.com/IQSS/dataverse/blob/a9605376c6be3d14c14592c1f692a3f83efae9df/src/main/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatter.java#L29-L60

### PR DESCRIPTION
**Pull request by alexscheitlin**
Hi guys,

as mentioned in #5821 I tested the function `SolrQueryFormatter.formatIdsForSolrClause`. I compiled a list of 45 test cases and implemented them using the `Parameterized` class of junit. There is a comment section in the respective test class on how the test cases are built.

The test cases reveal some room of improvement for the method under test if we look at the expected results and compare them with the corresponding inputs to the function. Below, you'll find some suggestions for adjusting the method.

## Related Issues

- fixes #5821: Unit Test `formatIdsForSolrClause` of `SolrQueryFormatter.java`

## Pull Request Checklist

- [x] Unit [tests][] completed
- [x] Integration [tests][]: None
- [x] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [x] [Documentation][docs] completed: None
- [x] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/test/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatterTest.java [hunk 1]
```diff
@@ -127,12 +133,103 @@ private void msgt(String s){
...
+                { new ArrayList<Long>(),                        "paramName", "dvObjectType", "sliceOfIds must have at least 1 value",           IllegalStateException.class },
+
+                { new ArrayList<Long>() {{ add(null); }},       null,        null,           "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>() {{ add(null); }},       null,        "",             "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>() {{ add(null); }},       null,        "dvObjectType", "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>() {{ add(null); }},       "",          null,           "(:())",                                           null },
+                { new ArrayList<Long>() {{ add(null); }},       "",          "",             "(:() AND dvObjectType:())",                       null },
+                { new ArrayList<Long>() {{ add(null); }},       "",          "dvObjectType", "(:() AND dvObjectType:(dvObjectType))",           null },
```
Comment (alexscheitlin) [inline src/test/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatterTest.java]: The expected query part seems not to make sense.
=> Check whether `paramName` is empty or not and throw an exception if it is null.

File: src/test/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatterTest.java [hunk 2]
```diff
@@ -127,12 +133,103 @@ private void msgt(String s){
...
+                { new ArrayList<Long>() {{ add(null); }},       "paramName", "",             "(paramName:() AND dvObjectType:())",              null },
+                { new ArrayList<Long>() {{ add(null); }},       "paramName", "dvObjectType", "(paramName:() AND dvObjectType:(dvObjectType))",  null },
+
+                { new ArrayList<Long>(Arrays.asList(1L)),       null,        null,           "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>(Arrays.asList(1L)),       null,        "",             "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>(Arrays.asList(1L)),       null,        "dvObjectType", "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>(Arrays.asList(1L)),       "",          null,           "(:(1))",                                          null },
+                { new ArrayList<Long>(Arrays.asList(1L)),       "",          "",             "(:(1) AND dvObjectType:())",                      null },
```
Comment (alexscheitlin) [inline src/test/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatterTest.java]: The expected query part seems not to make sense.
=> Check whether `dvObjectType` is empty or not and throw an exception if it is empty.

File: src/test/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatterTest.java [hunk 3]
```diff
@@ -127,12 +133,103 @@ private void msgt(String s){
...
+                { new ArrayList<Long>(),                        "paramName", "",             "sliceOfIds must have at least 1 value",           IllegalStateException.class },
+                { new ArrayList<Long>(),                        "paramName", "dvObjectType", "sliceOfIds must have at least 1 value",           IllegalStateException.class },
+
+                { new ArrayList<Long>() {{ add(null); }},       null,        null,           "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>() {{ add(null); }},       null,        "",             "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>() {{ add(null); }},       null,        "dvObjectType", "paramName cannot be null",                        NullPointerException.class },
+                { new ArrayList<Long>() {{ add(null); }},       "",          null,           "(:())",                                           null },
```
Comment (alexscheitlin) [inline src/test/java/edu/harvard/iq/dataverse/mydata/SolrQueryFormatterTest.java]: The expected query part seems not to make sense.
=> Check if there are only undefined ids (`null`) and throw an exception if there is no id defined as `Long`.

**Comment by kcondon** [comment]
@alexscheitlin Hi, would you mind refreshing this branch from /develop as the release version has changed? Thanks!

**Comment by alexscheitlin** [comment]
> @alexscheitlin Hi, would you mind refreshing this branch from /develop as the release version has changed? Thanks!

Hi @kcondon, sure. I merged the latest develop branch.

**Pull request merged**